# RAG Standalone Service

In [1]:
import requests

In [2]:
from requests.auth import HTTPBasicAuth
json_data = {
  "email": "ffm@example.com",
  "password": "passwordffm"
}
#response = requests.get(url, auth=HTTPBasicAuth(user, password))


In [3]:
url_login = 'https://api-656930476914.europe-west1.run.app/api/v1.0/user/login'

# Make the POST request
response = requests.post(url_login, json=json_data)

data = response.json()
print(response)


<Response [200]>


In [4]:
token = data['data']['id_token']

### General questions and KPI generation

In [5]:
url = 'https://api-656930476914.europe-west1.run.app/api/v1.0/kpi/'

headers = {
    'Authorization': f'Bearer {token}'
}

# Make the GET request
response = requests.get(url, headers=headers)

kpi_data = response.json()
print(response)
print(data)

<Response [422]>
{'success': True, 'data': {'uid': 'xM2kea8akaOKvYta26NMFBy8YnJ3', 'email': 'ffm@example.com', 'site': 1, 'first_name': 'Mario', 'last_name': 'Rossi', 'phone_number': '0987654321', 'id_token': 'eyJhbGciOiJSUzI1NiIsImtpZCI6IjFhYWMyNzEwOTkwNDljMGRmYzA1OGUwNjEyZjA4ZDA2YzMwYTA0MTUiLCJ0eXAiOiJKV1QifQ.eyJyb2xlIjoiRkZNIiwiaXNzIjoiaHR0cHM6Ly9zZWN1cmV0b2tlbi5nb29nbGUuY29tL3NtYXJ0YXBwLTlmMjg3IiwiYXVkIjoic21hcnRhcHAtOWYyODciLCJhdXRoX3RpbWUiOjE3MzQzNjc3NDcsInVzZXJfaWQiOiJ4TTJrZWE4YWthT0t2WXRhMjZOTUZCeThZbkozIiwic3ViIjoieE0ya2VhOGFrYU9Ldll0YTI2Tk1GQnk4WW5KMyIsImlhdCI6MTczNDM2Nzc0NywiZXhwIjoxNzM0MzcxMzQ3LCJlbWFpbCI6ImZmbUBleGFtcGxlLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwiZmlyZWJhc2UiOnsiaWRlbnRpdGllcyI6eyJlbWFpbCI6WyJmZm1AZXhhbXBsZS5jb20iXX0sInNpZ25faW5fcHJvdmlkZXIiOiJwYXNzd29yZCJ9fQ.ZJ8r41voZCOcVcDTJ5EUR3S670q-bs9naVs-36MSCk0OKqqnk7BzamJwQ605_vlsuBv2RjAw5wMKAOMSLvB49QCGqWbOZZp0IkvacnyKB-DLkd7M35TmwTgVY05zDX54o9fW-t1y_-ITwmHY4a4gl7V8UjCJ4i73w06vaySGSTFX2ZIf-JLFBT_vWXqURILyrN4qJx4ImVYy

In [6]:
machine_url = 'https://api-656930476914.europe-west1.run.app/api/v1.0/machine/'

# Make the GET request
response = requests.get(machine_url, headers=headers)

machine_data = response.json()
print(response)

<Response [200]>


In [7]:
print(machine_data)

{'success': True, 'data': [{'_id': '6740f1cfa8e3f95f42703128', 'category': 'Metal cutter', 'name': 'Large Capacity Cutting Machine 1', 'asset_id': 'ast-yhccl1zjue2t', 'kpis_ids': ['673a6ad2d9e0b151b88cbed0', '673a6ad2d9e0b151b88cbed1', '673a6ad3d9e0b151b88cbed2', '673a6ad3d9e0b151b88cbed3', '673a6ad4d9e0b151b88cbed4', '673a6ad4d9e0b151b88cbed5', '673a6ad4d9e0b151b88cbed6', '673a6ad5d9e0b151b88cbed7', '673a6ad5d9e0b151b88cbed8', '673a6ad5d9e0b151b88cbed9', '673a6ad6d9e0b151b88cbeda', '673a6ad7d9e0b151b88cbedb', '673a6ad8d9e0b151b88cbedc', '673a6ad8d9e0b151b88cbedd']}, {'_id': '6740f1cfa8e3f95f42703129', 'category': 'Rivetter', 'name': 'Riveting Machine', 'asset_id': 'ast-o8xtn5xa8y87', 'kpis_ids': ['673a6ad2d9e0b151b88cbed0', '673a6ad2d9e0b151b88cbed1', '673a6ad3d9e0b151b88cbed2', '673a6ad3d9e0b151b88cbed3', '673a6ad4d9e0b151b88cbed4', '673a6ad4d9e0b151b88cbed5', '673a6ad4d9e0b151b88cbed6', '673a6ad5d9e0b151b88cbed7', '673a6ad5d9e0b151b88cbed8', '673a6ad5d9e0b151b88cbed9', '673a6ad6d9e0

We build the list of data used as the knowledge base of the RAG such that it can anser general questions and kpi generation

In [8]:
prediction_url = 'https://api-656930476914.europe-west1.run.app/api/v1.0/chat/'

# Make the GET request
response = requests.get(prediction_url, headers=headers)

prediction_data = response.json()
print(response)

<Response [200]>


In [9]:
print(prediction_data)

{'success': True, 'data': {'cost_prediction': {'Testing Machines': '0.010329395319865256', 'Metal Cutting Machines': '0.04637815462858034', 'Laser Cutter': '0.0005160956082603602', 'Assembly Machines': '0.007193601533433455', 'Riveting Machine': '0.009630899430778293', 'Laser Welding Machines': '0.023551859302206688'}, 'utilization': {'Machine': {'11': 'Assembly Machine 1', '14': 'Laser Cutter', '13': 'Assembly Machine 2', '7': 'Laser Welding Machine 1', '3': 'Low Capacity Cutting Machine 1', '9': 'Riveting Machine', '0': 'Testing Machine 3', '10': 'Testing Machine 2', '12': 'Assembly Machine 3', '1': 'Laser Welding Machine 2', '4': 'Large Capacity Cutting Machine 2', '15': 'Large Capacity Cutting Machine 1', '6': 'Medium Capacity Cutting Machine 1', '2': 'Medium Capacity Cutting Machine 2', '5': 'Medium Capacity Cutting Machine 3', '8': 'Testing Machine 1'}, 'Average Utilization Rate': {'11': 0.7296168144672052, '14': 0.7133254505802838, '13': 0.6512304181921241, '7': 0.63181662966517

In [10]:
response_data_list = [kpi_data, machine_data, prediction_data]

In [11]:
import json 
from rag_service.rag_lib import rag_interaction

In [12]:
query = "How many machines are there?"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/3 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/8 [00:00<?, ?it/s]

In [12]:
response_data_list

[{'detail': [{'type': 'missing',
    'loc': ['query', 'site'],
    'msg': 'Field required',
    'input': None}]},
 {'success': True,
  'data': [{'_id': '6740f1cfa8e3f95f42703128',
    'category': 'Metal cutter',
    'name': 'Large Capacity Cutting Machine 1',
    'asset_id': 'ast-yhccl1zjue2t',
    'kpis_ids': ['673a6ad2d9e0b151b88cbed0',
     '673a6ad2d9e0b151b88cbed1',
     '673a6ad3d9e0b151b88cbed2',
     '673a6ad3d9e0b151b88cbed3',
     '673a6ad4d9e0b151b88cbed4',
     '673a6ad4d9e0b151b88cbed5',
     '673a6ad4d9e0b151b88cbed6',
     '673a6ad5d9e0b151b88cbed7',
     '673a6ad5d9e0b151b88cbed8',
     '673a6ad5d9e0b151b88cbed9',
     '673a6ad6d9e0b151b88cbeda',
     '673a6ad7d9e0b151b88cbedb',
     '673a6ad8d9e0b151b88cbedc',
     '673a6ad8d9e0b151b88cbedd']},
   {'_id': '6740f1cfa8e3f95f42703129',
    'category': 'Rivetter',
    'name': 'Riveting Machine',
    'asset_id': 'ast-o8xtn5xa8y87',
    'kpis_ids': ['673a6ad2d9e0b151b88cbed0',
     '673a6ad2d9e0b151b88cbed1',
     '673a6ad3d

In [13]:
print(response)

<Response [200]>


In [14]:
query = "What is the cost prediction analysis for the Laser Cutter?"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/3 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/8 [00:00<?, ?it/s]

In [15]:
print(response)

{
  "KPIs": [
    {
      "name": "Laser Cutter Cost Prediction",
      "type": "cost_prediction",
      "description": "Average daily cost of using a Laser Cutter",
      "unit_of_measure": "EUR/kWh",
      "formula": "0.0005160956082603602"
    }
  ]
}


In [23]:
query = "what's is the most efficient machine in terms of energy efficiency?"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/3 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/8 [00:00<?, ?it/s]

In [24]:
print(response)

{
  "KPIs": [
    {
      "name": "Energy Efficiency",
      "type": "ratio",
      "description": "The ratio of idle energy consumption to total energy consumption",
      "unit_of_measure": "",
      "formula": "{consumption_idle}/{(consumption_idle + consumption_working)}"
    }
  ]
}


In [18]:
query = "What can you do?"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/3 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/8 [00:00<?, ?it/s]

In [20]:
print(response)

{
  "message": "I'm sorry, but I cannot assist with generating a new KPI formula for the provided query.",
  "success": false
}


In [10]:
query = "What was the average consumption of the Testing machine 1?"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/2 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/7 [00:00<?, ?it/s]

In [11]:
print(response)

The average consumption for the specified machine is not available in our knowledge base. However, you can use KPIs related to consumption and cycles associated with the Testing Machine 1. Check the kpis_ids linked to this machine and explore any relevant KPI connected to consumption. Look into both asset_id and category of the machine as well for precise results.


In [ ]:
query = "Generate a KPI which measures consumption over cycles"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/2 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/7 [00:00<?, ?it/s]

In [13]:
print(response)

{
"KPIs": [
    {
      "name": "consuption_over_cycles",
      "type": "ratio",
      "description": "The consumption over all cycles",
      "unit_of_measure": "/cycle",
      "formula": "(consumption / cycles) * 1"
    }
  ]
}


In [14]:
query = "how many machines are there?"
response = rag_interaction(response_data_list, query)

Parsing nodes:   0%|          | 0/2 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/7 [00:00<?, ?it/s]

In [15]:
print(response)

There is not enough information available to determine the exact number of machines. Our knowledge base contains various types of cutting machines, each with its unique specifications and capacities. If you could provide more details or clarify which type of machine you are referring to, I would be happy to try and assist you further.


### Report generation

In [ ]:
mock_data = {
  "industry": "Manufacturing",
  "report": {
    "executive_summary": {
      "overview": "The manufacturing sector experienced a steady growth of 4.2% in Q3 2024, driven by increased demand in the automotive and electronics segments. However, supply chain disruptions and rising raw material costs continue to challenge profitability.",
      "highlights": [
        "Revenue grew by 4.2% compared to Q2 2024.",
        "Operational costs rose by 3.8%, primarily due to higher energy and material costs.",
        "Emerging markets contributed 35% of total revenue."
      ]
    },
    "key_performance_indicators": {
      "revenue_growth": "4.2%",
      "operating_margin": "12.4%",
      "supply_chain_efficiency": "87%",
      "energy_consumption_per_unit": "5.6 kWh",
      "employee_productivity": "27 units/employee/day"
    },
    "trends_and_observations": {
      "positive_trends": [
        "Increased adoption of automation boosted productivity by 7%.",
        "Green manufacturing initiatives reduced waste by 12% year-over-year."
      ],
      "negative_trends": [
        "Persistent supply chain disruptions caused delivery delays in 15% of orders.",
        "Material costs rose by an average of 5% across key inputs like steel and aluminum."
      ]
    },
    "recommendations": [
      {
        "area": "Supply Chain",
        "actions": [
          "Diversify suppliers to reduce dependency on regions prone to geopolitical risks.",
          "Implement advanced analytics for real-time supply chain monitoring."
        ]
      },
      {
        "area": "Cost Management",
        "actions": [
          "Negotiate long-term contracts with key suppliers to lock in material costs.",
          "Invest in energy-efficient technologies to lower operational costs."
        ]
      },
      {
        "area": "Product Innovation",
        "actions": [
          "Expand R&D efforts to develop high-demand sustainable products.",
          "Leverage customer feedback to refine existing product lines."
        ]
      }
    ],
    "data_appendix": {
      "quarterly_data": [
        {
          "quarter": "Q1 2024",
          "revenue": 45000000,
          "operating_margin": "13%",
          "supply_chain_efficiency": "85%",
          "energy_consumption_per_unit": "5.8 kWh"
        },
        {
          "quarter": "Q2 2024",
          "revenue": 47000000,
          "operating_margin": "12.8%",
          "supply_chain_efficiency": "86%",
          "energy_consumption_per_unit": "5.7 kWh"
        },
        {
          "quarter": "Q3 2024",
          "revenue": 49000000,
          "operating_margin": "12.4%",
          "supply_chain_efficiency": "87%",
          "energy_consumption_per_unit": "5.6 kWh"
        }
      ],
      "regional_revenue": [
        {
          "region": "North America",
          "revenue": 21000000,
          "growth_rate": "5%"
        },
        {
          "region": "Europe",
          "revenue": 18000000,
          "growth_rate": "3%"
        },
        {
          "region": "Asia-Pacific",
          "revenue": 10000000,
          "growth_rate": "6%"
        }
      ]
    }
  }
}

In [ ]:
import json 
from rag_service.rag_lib import rag_interaction

response = rag_interaction(mock_data)

Generating embeddings: 100%|██████████| 2/2 [00:02<00:00,  1.16s/it]


In [3]:
print(response)

**Executive Summary**

The analyzed data reveals a complex landscape of industry trends, opportunities, and challenges. Key findings include an increase in demand for sustainable practices and technological advancements in manufacturing processes. However, this growth comes with rising costs and logistical complexities.

**Key Performance Indicators (KPIs)**

| KPI | Current Value | Target Value | Variance |
| --- | --- | --- | --- |
| Revenue Growth Rate | 12% | 15% | -20% |
| Production Capacity Utilization | 80% | 85% | -5% |
| Energy Consumption (kWh) | 500,000 | 450,000 | +10% |

**Trends and Observations**

* The adoption of Industry 4.0 technologies is on the rise, driven by the need for increased efficiency and competitiveness.
* Supply chain disruptions remain a significant challenge, with 75% of respondents citing this as a major concern.
* There is growing demand for circular economy practices, with 60% of companies planning to increase their use of recycled materials.

**Re

**Executive Summary**

The analyzed data reveals a complex landscape of industry trends, opportunities, and challenges. Key findings include an increase in demand for sustainable practices and technological advancements in manufacturing processes. However, this growth comes with rising costs and logistical complexities.

**Key Performance Indicators (KPIs)**

| KPI | Current Value | Target Value | Variance |
| --- | --- | --- | --- |
| Revenue Growth Rate | 12% | 15% | -20% |
| Production Capacity Utilization | 80% | 85% | -5% |
| Energy Consumption (kWh) | 500,000 | 450,000 | +10% |

**Trends and Observations**

* The adoption of Industry 4.0 technologies is on the rise, driven by the need for increased efficiency and competitiveness.
* Supply chain disruptions remain a significant challenge, with 75% of respondents citing this as a major concern.
* There is growing demand for circular economy practices, with 60% of companies planning to increase their use of recycled materials.

**Recommendations**

1. **Invest in Industry 4.0 technologies**: Companies should prioritize the adoption of digital transformation strategies to enhance operational efficiency and stay competitive.
2. **Mitigate supply chain risks**: Diversify supplier networks and invest in logistics infrastructure to minimize disruptions and ensure timely delivery of critical components.
3. **Promote circular economy practices**: Develop strategic plans to increase the use of recycled materials, reduce waste, and improve sustainability.

**Data Appendix**

Additional data points used in this analysis:

| Category | Value |
| --- | --- |
| Average production cost per unit | $25.50 |
| Sales volume (units) | 100,000 |
| Employee headcount | 500 |

This report provides a comprehensive overview of the industry's current state and offers actionable insights for improvement. By addressing these key areas, companies can capitalize on emerging trends and drive long-term growth and success.

If using the actual data from the api

In [1]:
import requests

# Login URL and data
url_login = 'https://api-656930476914.europe-west1.run.app/api/v1.0/user/login'
login_data = {
    "email": "ffm@example.com",  # Replace with actual login email
    "password": "passwordffm"    # Replace with actual login password
}

# Perform login to get the token
login_response = requests.post(url_login, json=login_data)

if login_response.status_code == 200:
    login_data = login_response.json()
    token = login_data['data']['id_token']
    print("Login successful. Token acquired.")
else:
    print(f"Login failed. Status code: {login_response.status_code}, Response: {login_response.text}")
    exit()

# KPI API endpoint and parameters
url_kpi = 'https://api-656930476914.europe-west1.run.app/api/v1.0/kpi/site/1/report'
params = {
    'start_date': '2024-09-30 00:00:00',
    'end_date': '2024-10-01 00:00:00',
    'granularity_op': 'avg'
}

# Headers including the token
headers = {
    'accept': 'application/json',
    'Authorization': f'Bearer {token}'
}

# Make the GET request to the KPI endpoint
kpi_response = requests.get(url_kpi, headers=headers, params=params)

if kpi_response.status_code == 200:
    report_data = kpi_response.json()
    print("KPI data retrieved successfully.")
    print(report_data)
else:
    print(f"KPI request failed. Status code: {kpi_response.status_code}, Response: {kpi_response.text}")


Login successful. Token acquired.
KPI data retrieved successfully.
{'success': True, 'data': {'start_date': '2024-09-30T00:00:00', 'end_date': '2024-10-01T00:00:00', 'op': 'avg', 'kpis': [{'name': 'working_time', 'value': 27679.0}, {'name': 'idle_time', 'value': 14460.1}, {'name': 'offline_time', 'value': 846.5}, {'name': 'consumption', 'value': 0.00182641489881293}, {'name': 'power', 'value': 0.005303061508443911}, {'name': 'consumption_working', 'value': 0.0041813749405849405}, {'name': 'consumption_idle', 'value': 0.00099094737948828}, {'name': 'cost', 'value': 0.00129749009555441}, {'name': 'cost_working', 'value': 0.00139517124575466}, {'name': 'cost_idle', 'value': 0.0008199700085166799}, {'name': 'cycles', 'value': 1.0}, {'name': 'good_cycles', 'value': 1501.8}, {'name': 'bad_cycles', 'value': 6.4}, {'name': 'average_cycle_time', 'value': 6.5414536024002725}, {'name': 'production_cost_per_unit', 'value': 0.00129749009555441}, {'name': 'energy_consumption_per_unit', 'value': 0.00

In [2]:
data = [report_data['data']]

In [3]:
data

[{'start_date': '2024-09-30T00:00:00',
  'end_date': '2024-10-01T00:00:00',
  'op': 'avg',
  'kpis': [{'name': 'working_time', 'value': 27679.0},
   {'name': 'idle_time', 'value': 14460.1},
   {'name': 'offline_time', 'value': 846.5},
   {'name': 'consumption', 'value': 0.00182641489881293},
   {'name': 'power', 'value': 0.005303061508443911},
   {'name': 'consumption_working', 'value': 0.0041813749405849405},
   {'name': 'consumption_idle', 'value': 0.00099094737948828},
   {'name': 'cost', 'value': 0.00129749009555441},
   {'name': 'cost_working', 'value': 0.00139517124575466},
   {'name': 'cost_idle', 'value': 0.0008199700085166799},
   {'name': 'cycles', 'value': 1.0},
   {'name': 'good_cycles', 'value': 1501.8},
   {'name': 'bad_cycles', 'value': 6.4},
   {'name': 'average_cycle_time', 'value': 6.5414536024002725},
   {'name': 'production_cost_per_unit', 'value': 0.00129749009555441},
   {'name': 'energy_consumption_per_unit', 'value': 0.00182641489881293},
   {'name': 'power_effi

In [4]:
import json 
from rag_service.rag_lib import rag_interaction

response = rag_interaction(data)

/Users/nicolapitzalis/anaconda3/envs/rag/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Generating embeddings: 100%|██████████| 1/1 [00:02<00:00,  2.49s/it]


In [5]:
print(response)

Executive Summary
-----------------

The industrial data analyzed over the specified time period, from September 30, 2024, to October 1, 2024, indicates a relatively stable operational performance. However, there are some notable trends and observations that warrant attention.

Key Performance Indicators (KPIs)
-----------------------------------

### Production Metrics

| KPI | Value |
| --- | --- |
| Working Time | 27,679 hours |
| Idle Time | 14,460 hours |
| Offline Time | 846 hours |
| Production Cost per Unit | $0.01297 |

### Energy Consumption and Efficiency

| KPI | Value |
| --- | --- |
| Energy Consumption per Unit | 1.826 kJ/kWh |
| Power Efficiency | 809796,110 |
| Power Distribution Loss | -2,588.344 kW |

### Quality Metrics

| KPI | Value |
| --- | --- |
| Good Cycles | 1,501 |
| Bad Cycles | 6 |
| Average Cycle Time | 6.54 minutes |

Trends and Observations
-------------------------

* The working time has increased by approximately 10% compared to the previous period,

Executive Summary
-----------------

The industrial data analyzed over the specified time period, from September 30, 2024, to October 1, 2024, indicates a relatively stable operational performance. However, there are some notable trends and observations that warrant attention.

Key Performance Indicators (KPIs)
-----------------------------------

### Production Metrics

| KPI | Value |
| --- | --- |
| Working Time | 27,679 hours |
| Idle Time | 14,460 hours |
| Offline Time | 846 hours |
| Production Cost per Unit | $0.01297 |

### Energy Consumption and Efficiency

| KPI | Value |
| --- | --- |
| Energy Consumption per Unit | 1.826 kJ/kWh |
| Power Efficiency | 809796,110 |
| Power Distribution Loss | -2,588.344 kW |

### Quality Metrics

| KPI | Value |
| --- | --- |
| Good Cycles | 1,501 |
| Bad Cycles | 6 |
| Average Cycle Time | 6.54 minutes |

Trends and Observations
-------------------------

* The working time has increased by approximately 10% compared to the previous period, indicating a potential increase in production activity.
* Idle time has decreased by 13.5%, suggesting efforts to optimize downtime or improve process efficiency.
* Offline time remains relatively stable at around 846 hours, indicating that maintenance and repair activities are being managed effectively.

Recommendations
-----------------

Based on these findings, we recommend:

* Continuously monitoring working time and idle time to identify opportunities for improvement in production planning and scheduling.
* Implementing strategies to reduce downtime or optimize process efficiency to minimize offline time.
* Regularly reviewing energy consumption patterns to ensure that the power distribution system is functioning efficiently.

Data Appendix
--------------

The following table provides a summary of the KPIs extracted from the JSON data:

| Category | KPI Name | Value |
| --- | --- | --- |
| Production | Working Time | 27,679 hours |
| Production | Idle Time | 14,460 hours |
| Production | Offline Time | 846 hours |
| Energy Consumption | Energy Consumption per Unit | 1.826 kJ/kWh |
| Energy Efficiency | Power Efficiency | 809796,110 |
| Quality | Good Cycles | 1,501 |
| Quality | Bad Cycles | 6 |
| Quality | Average Cycle Time | 6.54 minutes |

By analyzing these KPIs and trends, we can gain insights into the performance of the industrial operation and make data-driven decisions to optimize efficiency and productivity.

Note: This report is generated based on the provided JSON data and is intended for informational purposes only.
